lib/src/layer/polyline_layer.dart

```diff
@@ -39,66 +79,292 @@ class Polyline {
+                    lastHit.hit!.tapped
+                        .map((e) {
+                          e.onSecondaryTap?.call(lastHit.hit!.point);
+                          return e.tapKey;
+                        })
+                        .whereNotNull()
+                        .toList(),
+                    lastHit.hit!.point,
+                  );
+                },
+                child: paint,
+              ),
+            )
+          : paint,
     );
   }
 }
 
-class PolylinePainter extends CustomPainter {
-  final List<Polyline> polylines;
-
-  final MapCamera map;
+class _PolylinePainter<TapKeyType extends Object> extends CustomPainter {
+  final List<Polyline<TapKeyType>> polylines;
+  final MapCamera camera;
   final LatLngBounds bounds;
+  final _LastHit? lastHit;
+  final double onTapTolerance;
+  final bool tappablesObscure;
+  final bool nonTappablesObscure;
 
-  PolylinePainter(this.polylines, this.map) : bounds = map.visibleBounds;
+  _PolylinePainter({
+    required this.polylines,
+    required this.camera,
+    required this.lastHit,
+    required this.onTapTolerance,
+    required this.tappablesObscure,
+    required this.nonTappablesObscure,
+  }) : bounds = camera.visibleBounds;
 
   int get hash => _hash ??= Object.hashAll(polylines);
 
   int? _hash;
 
+  @override
+  bool? hitTest(Offset position) {
+    if (lastHit == null) return null;
+
+    final hits = <Polyline<TapKeyType>>[];
+    final origin =
+        camera.project(camera.center).toOffset() - camera.size.toOffset() / 2;
+
+    outer:
+    for (final p in polylines.reversed) {
+      // TODO: For efficiency we'd ideally filter by bounding box here. However
+      // we'd need to compute an extended bounding box that accounts account for
+      // the stroke width.
+      // if (!p.boundingBox.contains(touch)) {
+      //   continue;
+      // }
+
+      final offsets = getOffsets(camera, origin, p.points);
+      final strokeWidth = p.useStrokeWidthInMeter
+          ? _metersToStrokeWidth(
+              origin,
+              p.points.first,
+              offsets.first,
+              p.strokeWidth,
+            )
+          : p.strokeWidth;
+      final maxDistance =
+          (strokeWidth / 2 + p.borderStrokeWidth / 2) + onTapTolerance;
+
+      for (int i = 0; i < offsets.length - 1; i++) {
+        final o1 = offsets[i];
+        final o2 = offsets[i + 1];
+
+        final distance = math.sqrt(_distToSegmentSquared(
+          position.dx,
+          position.dy,
+          o1.dx,
+          o1.dy,
+          o2.dx,
+          o2.dy,
+        ));
+
+        if (distance < maxDistance) {
+          if (nonTappablesObscure && p.onTap == null && p.tapKey == null) {
+            break outer;
+          }
+          hits.add(p);
+          if (tappablesObscure) break outer;
+        }
+      }
+    }
+
+    if (hits.isEmpty) return false;
+
+    lastHit!.hit = (
+      // Remove duplicates caused when the hit is close to two segements
```

Comment:
I don't understand this. Why would you end up with two segments of the same polyline? Why wouldn't you stop after you found the first hit segment?

Reply:
As we're handling multiple `Polyline`s now, and the detection algorithm seems to rely on comparing with each segment of the `Polyline`, a hit within two or more segments will trigger a result for the same line twice.

I guess this could be fixed by tweaking the detection algorithm to not depend on segments, but rather check the entire line, but I have no idea how to do that, and you're the expert there!

![image](https://github.com/fleaflet/flutter_map/assets/58115698/e3f9580a-12e2-441e-92a7-dc5da7d0a76f)

Reply:
> I guess this could be fixed by tweaking the detection algorithm to not depend on segments, but rather check the entire line, but I have no idea how to do that, and you're the expert there!

I don't know how you'd simply check the entire line instead of segments. However, as we iterate the segments sequentially and detect a hit, we stop and move on to the next polyline. No need to continue iterating segments for the same polyline.

Reply:
Ah, I've got you now. There was a bug where I wasn't `break`ing normally after one valid segment had been considered.